# Team assignment - Boston 2016

## Description of the dataset

In [33]:
import numpy as np #optional
import pandas as pd

### Bike rentals

Read data on bike rentals:

In [34]:
bike = pd.read_csv("boston_2016.csv") #test

In [22]:
#for later use
bike16 = pd.read_csv("boston_2016.csv")
bike17 = pd.read_csv("boston_2017.csv")
bike18 = pd.read_csv("boston_2018.csv")
bike19 = pd.read_csv("boston_2019.csv")

In [47]:
bike.head(20000)

start_time             end_time  start_station_id  \
0      2016-01-01 00:08:07  2016-01-01 00:40:26                36   
1      2016-01-01 00:08:19  2016-01-01 00:40:27                36   
2      2016-01-01 00:10:15  2016-01-01 00:40:29                36   
3      2016-01-01 00:10:22  2016-01-01 00:40:11                36   
4      2016-01-01 00:15:36  2016-01-01 00:31:17                36   
...                    ...                  ...               ...   
19995  2016-02-19 10:33:34  2016-02-19 10:36:09               116   
19996  2016-02-19 10:34:59  2016-02-19 10:40:46                87   
19997  2016-02-19 10:47:21  2016-02-19 11:10:38               140   
19998  2016-02-19 10:48:24  2016-02-19 10:59:44                74   
19999  2016-02-19 10:48:36  2016-02-19 10:54:46               115   

       end_station_id                                 start_station_name  \
0                  19           Boston Public Library - 700 Boylston St.   
1                  19           Boston Public Library - 700 Boylston St.   
2                  19           Boston Public Library - 700 Boylston St.   
3                  19           Boston Public Library - 700 Boylston St.   
4                  67           Boston Public Library - 700 Boylston St.   
...               ...                                                ...   
19995              76          359 Broadway - Broadway at Fayette Street   
19996             105  Harvard University Housing - 115 Putnam Ave at...   
19997              72                                        Danehy Park   
19998             107                Harvard Square at Mass Ave/ Dunster   
19999             104                              Porter Square Station   

                                        end_station_name  bike_id   user_type  
0                                Buswell St. at Park Dr.     1406  Subscriber  
1                                Buswell St. at Park Dr.     1121  Subscriber  
2                                Buswell St. at Park Dr.      474  Subscriber  
3                                Buswell St. at Park Dr.     1504  Subscriber  
4                           MIT at Mass Ave / Amherst St      913    Customer  
...                                                  ...      ...         ...  
19995  Central Sq Post Office / Cambridge City Hall a...      754  Subscriber  
19996    Lower Cambridgeport at Magazine St/Riverside Rd     1526  Subscriber  
19997      One Broadway / Kendall Sq at Main St / 3rd St      740  Subscriber  
19998                                 Ames St at Main St      935  Subscriber  
19999  Harvard University Radcliffe Quadrangle at She...      853  Subscriber  

[20000 rows x 8 columns]

In [36]:
bike.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236203 entries, 0 to 1236202
Data columns (total 8 columns):
 #   Column              Non-Null Count    Dtype 
---  ------              --------------    ----- 
 0   start_time          1236203 non-null  object
 1   end_time            1236203 non-null  object
 2   start_station_id    1236203 non-null  int64 
 3   end_station_id      1236203 non-null  int64 
 4   start_station_name  1236203 non-null  object
 5   end_station_name    1236203 non-null  object
 6   bike_id             1236203 non-null  int64 
 7   user_type           1236203 non-null  object
dtypes: int64(3), object(5)
memory usage: 75.5+ MB


Number of entries (1236203) = Non-Null count for all colums, meaning we are not missing any values.
Therefore not necessary to delete or imputate data/rows.

**Cleaning the datset:**

In [21]:
bike.isnull()

start_time  end_time  start_station_id  end_station_id  \
0             False     False             False           False   
1             False     False             False           False   
2             False     False             False           False   
3             False     False             False           False   
4             False     False             False           False   
...             ...       ...               ...             ...   
1236198       False     False             False           False   
1236199       False     False             False           False   
1236200       False     False             False           False   
1236201       False     False             False           False   
1236202       False     False             False           False   

         start_station_name  end_station_name  bike_id  user_type  
0                     False             False    False      False  
1                     False             False    False      False  
2                     False             False    False      False  
3                     False             False    False      False  
4                     False             False    False      False  
...                     ...               ...      ...        ...  
1236198               False             False    False      False  
1236199               False             False    False      False  
1236200               False             False    False      False  
1236201               False             False    False      False  
1236202               False             False    False      False  

[1236203 rows x 8 columns]

In [41]:
import datetime

weekday = weekday = list(map(lambda x : datetime.datetime.strptime(x, "%Y-%m-%d %H:%M:%S").weekday() <= 4, bike.end_time))
bike.loc[:,"IsWeekday"] = weekday

ValueError: unconverted data remains: .3210

In [25]:
bike.head(3)

start_time             end_time  start_station_id  end_station_id  \
0  2016-01-01 00:08:07  2016-01-01 00:40:26                36              19   
1  2016-01-01 00:08:19  2016-01-01 00:40:27                36              19   
2  2016-01-01 00:10:15  2016-01-01 00:40:29                36              19   

                         start_station_name         end_station_name  bike_id  \
0  Boston Public Library - 700 Boylston St.  Buswell St. at Park Dr.     1406   
1  Boston Public Library - 700 Boylston St.  Buswell St. at Park Dr.     1121   
2  Boston Public Library - 700 Boylston St.  Buswell St. at Park Dr.      474   

    user_type  
0  Subscriber  
1  Subscriber  
2  Subscriber

### Weather

Read weather data:

In [42]:
weatherhour = pd.read_csv("weather_hourly_boston.csv")

In [25]:
weatherhour.colums #prob. not possible because data type is not an (python) object

AttributeError: 'DataFrame' object has no attribute 'colums'

In [44]:
weatherhour.head(10)

date_time  max_temp  min_temp  precip
0  2015-01-02 01:00:00      -1.1      -1.1     0.0
1  2015-01-02 02:00:00      -1.1      -1.1     0.0
2  2015-01-02 03:00:00      -0.6      -0.6     0.0
3  2015-01-02 04:00:00      -0.6      -0.6     0.0
4  2015-01-02 05:00:00      -0.6      -0.6     0.0
5  2015-01-01 06:00:00      -5.6      -5.6     0.0
6  2015-01-01 07:00:00      -5.6      -5.6     0.0
7  2015-01-01 08:00:00      -5.6      -5.6     0.0
8  2015-01-01 09:00:00      -4.4      -4.4     0.0
9  2015-01-01 10:00:00      -5.6      -5.6     0.0

In [28]:
weatherhour.describe()

max_temp      min_temp        precip
count  43354.000000  43354.000000  43356.000000
mean      11.456717     11.395922      0.095581
std       10.118606     10.114095      0.294019
min      -22.800000    -22.800000      0.000000
25%        3.300000      3.300000      0.000000
50%       11.700000     11.700000      0.000000
75%       20.000000     19.850000      0.000000
max       36.100000     36.100000      1.000000

In [27]:
weatherhour.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43848 entries, 0 to 43847
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date_time  43354 non-null  object 
 1   max_temp   43354 non-null  float64
 2   min_temp   43354 non-null  float64
 3   precip     43356 non-null  float64
dtypes: float64(3), object(1)
memory usage: 1.3+ MB


Number of entries (43848) != Non-Null count for all colums, meaning we are (!) missing values.

**Cleaning the datset:**

In [30]:
weatherhour.isnull()

date_time  max_temp  min_temp  precip
0          False     False     False   False
1          False     False     False   False
2          False     False     False   False
3          False     False     False   False
4          False     False     False   False
...          ...       ...       ...     ...
43843      False     False     False   False
43844      False     False     False   False
43845      False     False     False   False
43846      False     False     False   False
43847      False     False     False   False

[43848 rows x 4 columns]

In [50]:
#display entries with NaN values
NaN_for_attribute = weatherhour[weatherhour["date_time"].isna()] #insert any attribute
NaN_for_attribute

date_time  max_temp  min_temp  precip
98          NaN       NaN       NaN     NaN
145         NaN       NaN       NaN     NaN
304         NaN       NaN       NaN     NaN
395         NaN       NaN       NaN     NaN
588         NaN       NaN       NaN     NaN
...         ...       ...       ...     ...
43019       NaN       NaN       NaN     NaN
43428       NaN       NaN       NaN     NaN
43570       NaN       NaN       NaN     NaN
43594       NaN       NaN       NaN     NaN
43669       NaN       NaN       NaN     NaN

[494 rows x 4 columns]

Mostly, for given entries values are missing for all attributes.

**Imputation** <br>
To maintain the completeness of the time series dataset, we have decided to impute missing values.